sql: support prepare INSERT statements

madelynnblue opened the pull request:
Make sure prepare statements are run in a transaction.

Use a new boolean flag on the planner to indicate prepare. Statement
implementations will gain knowledge of work to omit during prepare. An
alternative implementation would split up the prepare and exec portions of
the work. However, this would require passing around significant amounts
of state and I think the best implementation is the one here.

Keep a whitelist of statement types that support preparation instead of
attempting to exec them during prepare.

Fixes #3819

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/3958)

<!-- Reviewable:end -->

petermattis commented:
I'm not enthused by the `prepareOnly` approach. Feels fragile to have to sprinkle `if p.prepareOnly` in the code rather than have a distinct `prepare` phase (performed by `makePlan`) and then have all of the work done in the first call to `plan.Next`. Perhaps I'm worrying unnecessarily. 

It's unfortunate that we have to replicate so much work for prepared statements that are subsequently executed. We'll likely have to fix that at some point. 

---

Review status: 0 of 11 files reviewed at latest revision, 3 unresolved discussions.

---

*[sql/executor.go, line 163 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3958#-K8tzQdi4vOYpMGIrd6L:-K8tzQdi4vOYpMGIrd6M:434047532) ([raw file](https://github.com/cockroachdb/cockroach/blob/613f6d2b2a2128310ade9feacb19e0585d4fca53/sql/executor.go#L163)):*
Statement preparation should never mutate the database. Perhaps abort the transaction here. Even better, rather than using `DB.Txn` (which commits) you should manually construct a `client.Txn` here and never commit it. 

---

*[sql/pgwire_test.go, line 340 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3958#-K8txk9vMPVzAQqoP49g:-K8txk9vMPVzAQqoP49h:-721381561) ([raw file](https://github.com/cockroachdb/cockroach/blob/613f6d2b2a2128310ade9feacb19e0585d4fca53/sql/pgwire_test.go#L340)):*
I think it would be good to be more exhaustive in the statements that are tested here.

---

*[sql/plan.go, line 160 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3958#-K8tyn2xcEPJHSw5OEtA:-K8tyn2xcEPJHSw5OEtB:-1870851779) ([raw file](https://github.com/cockroachdb/cockroach/blob/613f6d2b2a2128310ade9feacb19e0585d4fca53/sql/plan.go#L160)):*
Is the plan to add support for `prepareOnly` for all statements?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3958#-:-K8u-Y5Py2spZ0_7M4Up:1523888340)*
<!-- Sent from Reviewable.io -->

mjibson commented:
I agree that the `prepareOnly` approach is not great. I do think it makes sense to do a later refactor that can both remove `prepareOnly` in favor of the 2-step approach and do the plan saving at the same time. How's that sound?

---

Review status: 0 of 11 files reviewed at latest revision, 3 unresolved discussions.

---

*[sql/executor.go, line 163 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3958#-K8tzQdi4vOYpMGIrd6L:-K8uA76FmGsS2oL5ScBw:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/613f6d2b2a2128310ade9feacb19e0585d4fca53/sql/executor.go#L163)):*
Done.

---

*[sql/pgwire_test.go, line 340 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3958#-K8txk9vMPVzAQqoP49g:-K8u5Pm0WwQ20ox0GwuL:-1898246791) ([raw file](https://github.com/cockroachdb/cockroach/blob/613f6d2b2a2128310ade9feacb19e0585d4fca53/sql/pgwire_test.go#L340)):*
Do you mean other kinds of statements (like UPDATE)? If so, they don't work yet and will be added in the future.

---

*[sql/plan.go, line 160 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3958#-K8tyn2xcEPJHSw5OEtA:-K8u5o5vFDHcQDgN_HJO:-1720952221) ([raw file](https://github.com/cockroachdb/cockroach/blob/613f6d2b2a2128310ade9feacb19e0585d4fca53/sql/plan.go#L160)):*
Not sure. UPDATE and DELETE are the only two other statement types that I can recall would benefit from param values, so they are on the todo list. However, we may want to add prepare support to all statement types. I'm not sure it makes sense to have a prepared CREATE or DROP, but maybe it does? We discussed setting the timestamp of the statement execution to the time of the preparation (to protect against later schema changes and other time-based things), so those statements (CREATE, DROP, etc.) pretty much only make sense to exec once.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3958#-:-K8uI0bX698her7wN682:-76165803)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Sounds reasonable to me. Let's add a `TODO` next to `planner.prepareOnly` to indicate it is only a stop-gap solution and to sketch out the longer term plan.

---

Review status: 0 of 11 files reviewed at latest revision, 2 unresolved discussions.

---

*[sql/pgwire_test.go, line 340 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3958#-K8txk9vMPVzAQqoP49g:-K8uQSvqnJBDkeuByS_z:1502923265) ([raw file](https://github.com/cockroachdb/cockroach/blob/613f6d2b2a2128310ade9feacb19e0585d4fca53/sql/pgwire_test.go#L340)):*
Yeah, I wrote this before I saw that only `INSERT` and `SELECT` were implemented. But I think we need to test all statement types here as they should all be valid for preparing. 

---

*[sql/plan.go, line 160 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3958#-K8tyn2xcEPJHSw5OEtA:-K8uPg7lLDZf5eSVbgVL:763755148) ([raw file](https://github.com/cockroachdb/cockroach/blob/613f6d2b2a2128310ade9feacb19e0585d4fca53/sql/plan.go#L160)):*
I don't think a prepared `CREATE` or `DROP` make much sense, but the SQL driver might reasonably prepare all statements and then execute them. I think @knz or somebody was seeing a prepare on something like `SHOW` or `SET`.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3958#-:-K8uQ6ju9ogbJpklAHYr:1155349959)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Review status: 0 of 11 files reviewed at latest revision, 1 unresolved discussion.

---

*[sql/pgwire_test.go, line 340 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3958#-K8txk9vMPVzAQqoP49g:-K8zZDH-AKehdzhce51_:-483572871) ([raw file](https://github.com/cockroachdb/cockroach/blob/613f6d2b2a2128310ade9feacb19e0585d4fca53/sql/pgwire_test.go#L340)):*
Will add those in a later PR.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3958)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Just talked to Matt a bit. Let's make the longer term plan a shorter term plan that I make type checking a "phase" for all statements/expressions, which will do several things - type the numeric literals and infer types for valArgs.

---

Review status: 0 of 11 files reviewed at latest revision, 2 unresolved discussions.

---

*[sql/plan.go, line 36 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3958#-K8zOf-W0aXXO6D2f7_Q:-K8zOf-Xnzmp0ej2az8_:1714580955) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8339d94f71806db052af67da8995469b952d20a/sql/plan.go#L36)):*
please add a comment about this field

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3958#-:-K8zYRruTIFh2VSh99pX:417889805)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Review status: 0 of 11 files reviewed at latest revision, 2 unresolved discussions.

---

*[sql/plan.go, line 36 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3958#-K8zOf-W0aXXO6D2f7_Q:-K9--KGbzmimbpeHTpWM:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8339d94f71806db052af67da8995469b952d20a/sql/plan.go#L36)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3958)*
<!-- Sent from Reviewable.io -->

madelynnblue committed:
sql: support prepare INSERT statements

Make sure prepare statements are run in a transaction.

Use a new boolean flag on the planner to indicate prepare. Statement
implementations will gain knowledge of work to omit during prepare. An
alternative implementation would split up the prepare and exec portions of
the work. However, this would require passing around significant amounts
of state and I think the best implementation is the one here.

Keep a whitelist of statement types that support preparation instead of
attempting to exec them during prepare.

Fixes #3819

mjibson commented:
Closing in favor of #4037
